collections/_posts/2023-12-20-CoCs.md

```diff
@@ -0,0 +1,121 @@
+<img src="/collections/images/CoCs/centers_features.jpg" width=200></div>
+<p style="text-align: center;font-style:italic">Figure 3. Illustration of the computation of the center features. The red squares are the centers, the blue circle show the feature that are averaged for the top center, and the color palette shows the clusters of points associated with each centers.</p>
+
+&nbsp;
+
+**Feature aggregating.** &ensp; Given a cluster $$c$$ with $$m$$ points, the points are mapped to a high-dimension *value space* $$\textbf{P}_v \in \mathbb{R}^{m \times d'}$$. A center $$v_c$$ is also computed in the value space like it is done at the clustering step. 
+
+> In practice, the dimension of the value space is set to be equal to d' so this step can be skipped.
+
+The aggregated feature $$g \in \mathbb{R}^{d'}$$ of the cluster $$c$$  is computed as : 
+
+$$ g = \frac{1}{C} \left( v_c +  \sum_{i=1}^{m} \text{sig}(\alpha s_i + \beta) \ast v_i \right)$$
+
+with $$C = 1 + \sum_{i=1}^m \text{sig} (\alpha s_i \ast \beta)$$.
+
+$$v_i$$ is th $$i$$-th point in $$\textbf{P}_v$$, $$\alpha$$ and $$\beta$$ are learnt scalars and $$\text{sig}(\cdot)$$ is the sigmoïd function. $$v_c$$ and $$C$$ are added for numerical stability.
+
+**Feature dispatching.** &ensp; Once the aggregated feature $$g$$ has been computed inside a cluster, it is adaptively dispatched to each point in a cluster based on the similarity. This allows pixels inside a cluster to interact. Each point $$p_i$$ is updated as follows :
+
+$$ p_i ' = p_i + \text{FC}(\text{sig} (\alpha s_i + \beta) \ast g )$$
+
+where $$\text{FC}$$ is a fully-connected layer to get back to the original dimension $$d$$ from the value-space dimension $$d'$$
+
+A schematic view of a context cluster block is given is Figure 4.
+
+<div style="text-align:center">
+<img src="/collections/images/CoCs/cocs_block.jpg" width=700></div>
+<p style="text-align: center;font-style:italic">Figure 4. Overview of a Context Cluster block.</p>
+
+&nbsp;
+
+## Choices for computation time improving
+
+**Multi-head design.** &ensp; Like in ViTs, the authos use a multi-head design in the context cluster. The output of all the heads are concatenated and fused by a fully-connected layer.
+
+**Region splitting.** &ensp; If no restriction is set when associating the points with the centers, the time complexity to calculate the feature similarity is $$O(ncd)$$ (with $$n = h \times w$$), which is quadratic relatively to the number of points in the image. As it is too computationally demanding, the same technique as in SwinTransformer is done, that is to say **splitting the points into several local regions.**. If the number of regions is set to $$r$$, the time complexity is lowered to $$O(r\frac{n}{r}\frac{c}{r}d)$$. This obviously results in a trade-off between time complexity gain and limitation of the receptive field. In practice, the parameter $$r$$ is set so that the clusters are computed in a $$7 \times 7$$ window.
```

Comment:
Phrasing:
```suggestion
**Region splitting.** &ensp; If no restriction is set when associating the points with the centers, the time complexity to calculate the feature similarity is $$O(ncd)$$ (with $$n = h \times w$$), which is quadratic relative to the number of points in the image. As it is too computationally demanding, the same technique as in SwinTransformer is done, that is to say **splitting the points into several local regions.**. If the number of regions is set to $$r$$, the time complexity is lowered to $$O(r\frac{n}{r}\frac{c}{r}d)$$. This obviously results in a trade-off between time complexity gain and limitation of the receptive field. In practice, the parameter $$r$$ is set so that the clusters are computed in a $$7 \times 7$$ window.
```